Rover/mode_circle.cpp

```diff
@@ -146,13 +149,21 @@ void ModeCircle::update()
 
     // Update distance to destination and distance to edge
     const Vector2f center_to_veh = curr_pos - config.center_pos;
-    _distance_to_destination = center_to_veh.length();
-    dist_to_edge_m = fabsf(_distance_to_destination - config.radius);
+    _distance_to_destination = (target.pos.tofloat() - curr_pos).length();
+    dist_to_edge_m = fabsf(center_to_veh.length() - config.radius);
 
     // Update depending on stage
     if (!reached_edge) {
         update_drive_to_radius();
-
+    } else if (dist_to_edge_m > config.radius * 0.2 && !tracking_back) {
+        // if more than 20% inside or outside circle radius, slow vehicle
+        config.speed = 0.8 * config.speed;
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "Circle: Vehicle unable to turn fast enough to track circle. Slowing vehicle to %.1fm/s", config.speed);
+        tracking_back = true;
+    } else if (dist_to_edge_m < config.radius * 0.05 && tracking_back) {
```

Comment:
Would it make scene to use the same threshold as we do for starting the circle in the first place?

https://github.com/ArduPilot/ardupilot/blob/1e8e250459e9a68571f1c5265066b89fde7c5f4a/Rover/mode_circle.cpp#L174

Reply:
Sure, that'll work. Changed to use ``MAX(g2.turn_radius, AR_CIRCLE_REACHED_EDGE_DIST)`` instead